File: sdk/python/alp_sdk/tenant_mesh.py
from __future__ import annotations

from datetime import datetime, timezone
from typing import Dict, Any, Optional

class TenantMesh:
    def __init__(
        self,
        tenant_id: str,
        org_name: str,
        sso_provider: str = "SAML-Okta",
        max_workspaces: int = 50,
        created_at: Optional[str] = None,
    ):
        self.id = f"mesh-{tenant_id}"
        self.tenant_id = tenant_id
        self.org_name = org_name
        self.sso_provider = sso_provider
        self.max_workspaces = max_workspaces
        self.rbac_roles: Dict[str, str] = {}
        self.created_at = created_at or datetime.now(timezone.utc).isoformat()

class TenantMeshEngine:
    def __init__(self):
        self.tenants: Dict[str, TenantMesh] = {}

    def create_tenant_mesh(
        self,
        tenant_id: str,
        org_name: str,
        sso_provider: str = "SAML-Okta",
        max_workspaces: int = 50,
    ) -> TenantMesh:
        mesh = TenantMesh(
            tenant_id=tenant_id,
            org_name=org_name,
            sso_provider=sso_provider,
            max_workspaces=max_workspaces,
        )
        self.tenants[tenant_id] = mesh
        return mesh

    def assign_role(self, tenant_id: str, user_or_agent: str, role: str) -> bool:
        tenant = self.tenants.get(tenant_id)
        if not tenant:
            return False
        tenant.rbac_roles[user_or_agent] = role
        return True

    def authorize_action(self, tenant_id: str, user_or_agent: str, action: str) -> bool:
        tenant = self.tenants.get(tenant_id)
        if not tenant:
            return False

        role = tenant.rbac_roles.get(user_or_agent)
        if not role:
            return False

        if role == "admin":
            return True
        if role == "developer" and not action.startswith("tenant:delete"):
            return True
        if role == "auditor" and (action.startswith("audit:") or action.startswith("read:")):
            return True

        return False

File: parser/src/tenant-mesh.ts
export type EnterpriseRole = 'admin' | 'auditor' | 'developer';

export interface TenantMesh {
  id: string;
  tenantId: string;
  orgName: string;
  ssoProvider: string;
  maxWorkspaces: number;
  rbacRoles: Record<string, EnterpriseRole>;
  createdAt: string;
}

export class TenantMeshEngine {
  private tenants: Map<string, TenantMesh> = new Map();

  public createTenantMesh(
    tenantId: string,
    orgName: string,
    ssoProvider: string = 'SAML-Okta',
    maxWorkspaces: number = 50
  ): TenantMesh {
    const mesh: TenantMesh = {
      id: `mesh-${tenantId}`,
      tenantId,
      orgName,
      ssoProvider,
      maxWorkspaces,
      rbacRoles: {},
      createdAt: new Date().toISOString(),
    };

    this.tenants.set(tenantId, mesh);
    return mesh;
  }

  public assignRole(tenantId: string, userOrAgent: string, role: EnterpriseRole): boolean {
    const tenant = this.tenants.get(tenantId);
    if (!tenant) return false;
    tenant.rbacRoles[userOrAgent] = role;
    return true;
  }

  public authorizeAction(tenantId: string, userOrAgent: string, action: string): boolean {
    const tenant = this.tenants.get(tenantId);
    if (!tenant) return false;

    const role = tenant.rbacRoles[userOrAgent];
    if (!role) return false;

    if (role === 'admin') return true;
    if (role === 'developer' && !action.startsWith('tenant:delete')) return true;
    if (role === 'auditor' && (action.startsWith('audit:') || action.startsWith('read:'))) return true;

    return false;
  }

  public getTenant(tenantId: string): TenantMesh | undefined {
    return this.tenants.get(tenantId);
  }
}
